File: Week1/crack.c
#define _XOPEN_SOURCE
#include <unistd.h>

#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>

int main(int argc, char* argv[])
{
    // check for 2 arguments only
    if (argc != 2)
    {
        printf("Usage: ./crack hash\n");
        return 1;
    }


    // Hash password
    char* hash_pass = argv[1];
    // Salt string
    char salt[3];
    salt[0] = hash_pass[0];
    salt[1] = hash_pass[1];
    salt[2] = '\0';

     // Empty string for storing max 5 characters.
    char plaintext[5] = "";



    char* alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    int alpha_len = strlen(alphabet);

    // Variable for founded password
    int found = 0;


// Bruteforce

// 1 char pass

    for (int first = 0; first < alpha_len; first++)
    {
        plaintext[0] = alphabet[first];
        char* crypted_pass = crypt(plaintext, salt);
        if (strcmp(hash_pass, crypted_pass) == 0)
        {
            printf("%s\n", plaintext);
            found = 1;
        }

    }


// 2 char pass

    if (found == 0)
    {
        for (int first = 0; first < alpha_len; first++)
        {
            plaintext[0] = alphabet[first];

            for (int second = 0; second < alpha_len; second++)
                {
                    plaintext[1] = alphabet[second];
                    char* crypted_pass = crypt(plaintext, salt);

                    if (strcmp(hash_pass, crypted_pass) == 0)
                    {
                        printf("%s\n", plaintext);
                        found = 1;
                    }
                }
            }
        }


// 3 char pass
    if (found == 0)
    {
        for (int first = 0; first < alpha_len; first++)
        {
            plaintext[0] = alphabet[first];

            for (int second = 0; second < alpha_len; second++)
                {
                    plaintext[1] = alphabet[second];

                    for(int third = 0; third < alpha_len; third++)
                    {
                        plaintext[2] = alphabet[third];
                        char* crypted_pass = crypt(plaintext, salt);

                        if (strcmp(hash_pass, crypted_pass) == 0)
                        {
                            printf("%s\n", plaintext);
                            found = 1;
                        }
                    }
                }
        }
    }


// 4 char pass

    if (found == 0)
    {
        for (int first = 0; first < alpha_len; first++)
        {
            plaintext[0] = alphabet[first];

            for (int second = 0; second < alpha_len; second++)
                {
                    plaintext[1] = alphabet[second];

                    for(int third = 0; third < alpha_len; third++)
                    {
                        plaintext[2] = alphabet[third];

                        for(int forth = 0; forth < alpha_len; forth++)
                        {
                            plaintext[3] = alphabet[forth];
                            char* crypted_pass = crypt(plaintext, salt);

                            if (strcmp(hash_pass, crypted_pass) == 0)
                            {
                                printf("%s\n", plaintext);
                                found = 1;
                            }
                        }
                    }
                }
        }
    }



// 5 char pass
    if (found == 0)
    {
        for (int first = 0; first < alpha_len; first++)
        {
            plaintext[0] = alphabet[first];

            for (int second = 0; second < alpha_len; second++)
                {
                    plaintext[1] = alphabet[second];

                    for(int third = 0; third < alpha_len; third++)
                    {
                        plaintext[2] = alphabet[third];

                        for(int forth = 0; forth < alpha_len; forth++)
                        {
                            plaintext[3] = alphabet[forth];
                            for(int fifth = 0; fifth < alpha_len; fifth++)
                            {
                                plaintext[4] = alphabet[fifth];
                                char* crypted_pass = crypt(plaintext, salt);

                                if (strcmp(hash_pass, crypted_pass) == 0)
                                {
                                    printf("%s\n", plaintext);
                                    found = 1;
                                }
                            }
                        }
                    }
                }
        }
    }

    return 0;
}

File: Week7/crack.py
# A program that cracks passwords

from sys import argv, exit
from crypt import crypt
import string

# Check for correct number of command line arguments
if len(argv) != 2:
    print('Usage: python crack.py hash')
    exit(1)
hash_pass = argv[1]
salt = hash_pass[0:2]
letters = string.ascii_letters
plaintext = ''

# Bruteforce

# 1 char pass

for first in range(0, len(letters)):
    plaintext = letters[first]
    crypted_pass = crypt(plaintext, salt)
    if hash_pass == crypted_pass:
        print(plaintext)
        exit(1)

# 2 char pass
for first in range(0, len(letters)):
    for second in range(0, len(letters)):
        plaintext = letters[first] + letters[second]
        crypted_pass = crypt(plaintext, salt)
        if hash_pass == crypted_pass:
            print(plaintext)
            exit(1)

# 3 char pass
for first in range(0, len(letters)):
    for second in range(0, len(letters)):
        for third in range(0, len(letters)):
            plaintext = letters[first] + letters[second] + letters[third]
            crypted_pass = crypt(plaintext, salt)
            if hash_pass == crypted_pass:
                print(plaintext)
                exit(1)

# 4 char pass
for first in range(0, len(letters)):
    for second in range(0, len(letters)):
        for third in range(0, len(letters)):
            for forth in range(0, len(letters)):
                plaintext = letters[first] + letters[second] + letters[third] + letters[forth]
                crypted_pass = crypt(plaintext, salt)
                if hash_pass == crypted_pass:
                    print(plaintext)
                    exit(1)

# 5 char pass
for first in range(0, len(letters)):
    for second in range(0, len(letters)):
        for third in range(0, len(letters)):
            for forth in range(0, len(letters)):
                for fifth in range(0, len(letters)):
                    plaintext = letters[first] + letters[second] + letters[third] + letters[forth] + letters[fifth]
                    crypted_pass = crypt(plaintext, salt)
                    if hash_pass == crypted_pass:
                        print(plaintext)
                        exit(1)
